### PR TITLE
Chrome Debug Fails to Start #588

### PR DESCRIPTION
**Pull request by vrubezhny**
When Chrome/Firefox Debug launch is created, the only browser defined in 'Run with:` field of launch configuration is to be used to start debuging. So, no any additional searches for a browser executable to be performed.
If no browser is confiured, a message is displayed prompting a user to open Launch Configuration Browser Tab or Web Browser preference page to configure/select a correct browser.
    
Firefox Debug is also updated to serach fora browser executable in a list of browsers defined in Web Browser preference page.
If no executable can be found it is suggested to open Web Browser preference page to add/edit Firefox browser location.

Fixes: #588
Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### CODE REVIEW

**Comment by mickaelistria** [comment]
Code looks good.
I'm wondering whether instead of putting the logic in the LaunchAdapter, we should instead put this logic at bundle startup in order to update the known list of Web Browser directly in the web browser registry, and use the 1st one of them in case of default. What do you think? (note it can be a further iteration, no need to change it as part of this patch unless you think it's worth it)
Can you please bump version of org.eclipse.wildwebdeveloper (and other necessary artifacts) by +0.0.1

**Comment by akurtakov** [comment]
Should https://git.eclipse.org/c/platform/eclipse.platform.ui.git/tree/bundles/org.eclipse.ui.browser/plugin.xml#n164 be changed to be able to find more variations of chrome installs ?

**Comment by mickaelistria** [comment]
> Should https://git.eclipse.org/c/platform/eclipse.platform.ui.git/tree/bundles/org.eclipse.ui.browser/plugin.xml#n164 be changed to be able to find more variations of chrome installs ?

Are those browsers supposed to be added automatically or should user have to do something for them to be discovered? I can't find my firefox configured by default for instance.
I think it would indeed be great if we remove some logic from Wild Web Developer to reuse what's in the browser bundle.

**Comment by mickaelistria** [comment]
Thanks to Alex's hint, I think we should rely more on the browser plugin and even remove the lookup for chrome executables and instead just use the 1st declared browser that matches as default.

**Comment by vrubezhny** [comment]
I agree on that we shouldn't make any browser lookups here, but to use one of the matching  browsers defined on the `Web Browser`preference page. 
The problem is that `Browser` tab is marked as  `OK, No problem` even if no browser is selected as well as in case if a browser selection is done, but path is wrong/empty.
So a user still can try to launch the debugging session (and have the problems described in the issue).

**Comment by mickaelistria** [comment]
So let's have the tab validation turn the tab incomplete if no browser is selected.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		if (suggestEditLaunch) {
+			buttons.add(Messages.RuntimeExecutable_NotDefinedError_OpenLaunchConfigurationBrowserTab);
+		}
+		buttons.add(Messages.RuntimeExecutable_NotDefinedError_OpenWebBrowserPreferencePage);
+		buttons.add(IDialogConstants.CANCEL_LABEL);
+		
+		int result[] = {Window.CANCEL};
+		Display.getDefault().syncExec(() -> {
+			int response = suggestEditLaunch ?
+					MessageDialog.open(MessageDialog.ERROR, Activator.getShell(), title, message, SWT.NONE,
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java:66]: The shell should be a method parameter as with current implementation, using multiple screens, the dialog can show on the other monitor.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+@SuppressWarnings("restriction")
+public class MessageUtils {
+	private static final String BROWSER_TAB_NAME = "Browser"; //$NON-NLS-1$
+	private static final String WEB_BROWSER_PREFERENCE_PAGE = "org.eclipse.ui.browser.preferencePage"; //$NON-NLS-1$
+	
+	private MessageUtils() {
+		// stateless, don't instantiate
+	}
+	
+	public static int showBrowserLocationsConfigurationError(ILaunchConfiguration configuration, String mode, String browserName, boolean suggestEditLaunch) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java:48]: Can you please add a minimal comment to describe what's the return value?

**Review by mickaelistria — COMMENTED** [review]
Please see minor comments inline.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java (hunk 1 above)
Comment (vrubezhny) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java:66]: OK

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java (hunk 2 above)
Comment (vrubezhny) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java:48]: Sure

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+									IDialogConstants.CANCEL_LABEL);
+			if (!suggestEditLaunch) {
+				response++;
+			}
+			if (response == 0) {
+				ILaunchGroup group = DebugUITools.getLaunchGroup(configuration, mode);
+				if (group != null) {
+					LaunchGroupExtension groupExt = DebugUIPlugin.getDefault().getLaunchConfigurationManager().getLaunchGroup(group.getIdentifier());
+					if (groupExt != null) {
+						final LaunchConfigurationEditDialog dialog = new LaunchConfigurationEditDialog(Activator.getShell(), configuration, groupExt, false);
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java:85]: parentShell

Comment (vrubezhny) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/MessageUtils.java]: Yes. Sorry, forget about this one

**Review by mickaelistria — APPROVED** [review]
+1 to merge this good change as soon as Jenkins validates it.
It's a very good improvement from code and UX perspective, relying solely on upstream standard configuration!

**Comment by mickaelistria** [comment]
Build on Jenkins is passing well this time, let's merge it now and we'll deal with the occasional mirror issue later.

**Pull request merged**